src/gui/qgsfieldmappingmodel.cpp

```diff
@@ -0,0 +1,356 @@
+        {
+          return tr( "Type" );
+        }
+        case ColumnDataIndex::DestinationLength:
+        {
+          return tr( "Length" );
+        }
+        case ColumnDataIndex::DestinationPrecision:
+        {
+          return tr( "Precision" );
+        }
+      }
+    }
+    else if ( orientation == Qt::Vertical )
+    {
+      return section;
+    }
+  }
+  return QVariant();
+}
+
+QgsFields QgsFieldMappingModel::sourceFields() const
+{
+  return mSourceFields;
+}
+
+int QgsFieldMappingModel::rowCount( const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent );
+  return mMapping.count();
+}
+
+int QgsFieldMappingModel::columnCount( const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent );
+  return 5;
+}
+
+QVariant QgsFieldMappingModel::data( const QModelIndex &index, int role ) const
+{
+  if ( index.isValid() )
+  {
+    const int col { index.column() };
+    const Field &f { mMapping.at( index.row() ) };
+
+    if ( role == Qt::DisplayRole || role == Qt::EditRole )
+    {
+      switch ( col )
+      {
+          {
+          case ColumnDataIndex::SourceExpression:
+          {
+            return f.expression.expression();
+          }
+          case ColumnDataIndex::DestinationName:
+          {
+            return f.field.displayName();
+          }
+          case ColumnDataIndex::DestinationType:
+          {
+            return static_cast<int>( f.field.type() );
+          }
+          case ColumnDataIndex::DestinationLength:
+          {
+            return f.field.length();
+          }
+          case ColumnDataIndex::DestinationPrecision:
+          {
+            return f.field.precision();
+          }
+        }
+      }
+    }
+  }
+  return QVariant();
+}
+
+QgsExpressionContextGenerator *QgsFieldMappingModel::contextGenerator() const
+{
+  return mExpressionContextGenerator.get();
+}
+
+
+QgsFieldMappingModel::ExpressionContextGenerator::ExpressionContextGenerator( const QgsFields *sourceFields )
+{
+  mSourceFields = sourceFields;
+}
+
+QgsExpressionContext QgsFieldMappingModel::ExpressionContextGenerator::createExpressionContext() const
+{
+  QgsExpressionContext ctx;
+  ctx.appendScope( QgsExpressionContextUtils::globalScope() );
+  ctx.setFields( *mSourceFields );
+  QgsFeature feature { *mSourceFields };
+  feature.setValid( true );
+  ctx.setFeature( feature );
+  return ctx;
+}
+
+Qt::ItemFlags QgsFieldMappingModel::flags( const QModelIndex &index ) const
+{
+  if ( index.isValid() &&
+       ( index.column() == ColumnDataIndex::SourceExpression || destinationEditable() ) )
+  {
+    return Qt::ItemFlags( Qt::ItemIsSelectable |
+                          Qt::ItemIsEditable |
+                          Qt::ItemIsEnabled );
+  }
+  return Qt::ItemFlags();
+}
+
+bool QgsFieldMappingModel::setData( const QModelIndex &index, const QVariant &value, int role )
+{
+  if ( index.isValid() )
+  {
+    if ( role == Qt::EditRole )
+    {
+      Field &f { const_cast<Field &>( mMapping.at( index.row() ) ) };
```

Comment:
```suggestion
      Field &f = mMapping[index.row()];
```

Just for ease of readability :)

Reply:
> One thing, I just noticed this PR after merging #31387 which also aimed to make refactor fields easier to use. Do you see it feasible to integrate this in here too?

Yes, it shouldn't be hard.